### PR TITLE
Update code generation to GridTools v1.0.1

### PR DESCRIPTION
**Pull request by cosunae**

### CODE REVIEW

**Comment by havogt** [comment]
launch pipeline

File: src/dawn/IIR/IIR.cpp [hunk 1]
```diff
@@ -25,16 +27,64 @@
...
+    mergeFields(stencil->getFields(), derivedInfo_.fields_);
+  }
+}
+
+void IIR::DerivedInfo::clear() { fields_.clear(); }
+
 json::json IIR::jsonDump() const {
   json::json node;
 
+  json::json fieldsJson;
```
Comment (havogt) [inline src/dawn/IIR/IIR.cpp]: What the heck is that?

**Comment by havogt** [comment]
launch pipeline

**Comment by havogt** [comment]
launch pipeline

**Comment by havogt** [comment]
launch pipeline

**Comment by havogt** [comment]
launch pipeline

**Comment by havogt** [comment]
launch jenkins

**Comment by havogt** [comment]
launch jenkins

File: src/dawn/CodeGen/GridTools/GTCodeGen.cpp [hunk 2]
```diff
@@ -463,14 +518,14 @@ void GTCodeGen::generateStencilClasses(
                                                        intervalDefinitions.Levels.find(level)));
       int gt_offset =
           (level != sir::Interval::End) ? offset + 1 : (offset <= 0) ? offset - 1 : offset;
-      tss << "gridtools::level<" << gt_level << ", " << gt_offset << ", 4>";
+      tss << "gridtools::level<" << gt_level << ", " << gt_offset << ", 3>";
```
Comment (lukasm91) [inline src/dawn/CodeGen/GridTools/GTCodeGen.cpp]: I'm not sure about this 3. That should be the offset_limit of axis, which is 2 by default. But I am not 100% sure if this is also set here somewhere. I didn't find it yet.

**Comment by havogt** [comment]
launch jenkins

File: src/dawn/IIR/StencilInstantiation.cpp [hunk 3]
```diff
@@ -98,6 +98,18 @@ bool StencilInstantiation::isIDAccessedMultipleStencils(int accessID) const {
   return false;
 }
 
+bool StencilInstantiation::isIDAccessedMultipleMSs(int accessID) const {
+
+  int count = 0;
+  for(const auto& ms : iterateIIROver<MultiStage>(*IIR_)) {
+    if(ms->hasField(accessID)) {
+      if(++count > 1)
```
Comment (lukasm91) [inline src/dawn/IIR/StencilInstantiation.cpp]: this  is a weird function. Isn't this the same as just
```
if(ms->hasField(accessID)) return true;
```

File: src/dawn/Support/IndexRange.h [hunk 4]
```diff
@@ -49,34 +50,72 @@ struct IndexRangeIterator {
   using iterator =
       typename std::conditional<std::is_const<Cont>::value, typename Cont::const_iterator,
                                 typename Cont::iterator>::type;
+  using difference_type = typename iterator::difference_type;
+
+  // TODO make this private
```
Comment (lukasm91) [inline src/dawn/Support/IndexRange.h]: can we resolve this TODO?

File: src/dawn/CodeGen/CXXUtil.h [hunk 5]
```diff
@@ -521,8 +521,9 @@ struct Structure : public Statement {
   ///   template< templateName >
   ///   Structure(...) {...}
   /// @endcode
-  MemberFunction addDestructor() {
-    return addMemberFunction(Twine::createNull(), Twine("~") + StructureName);
+  MemberFunction addDestructor(bool isvirtual) {
```
Comment (lukasm91) [inline src/dawn/CodeGen/CXXUtil.h]:
```suggestion
  MemberFunction addDestructor(bool isVirtual) {
```

File: src/dawn/CodeGen/CXXUtil.h [hunk 6]
```diff
@@ -521,8 +521,9 @@ struct Structure : public Statement {
...
+    return addMemberFunction(Twine::createNull(),
+                             (isvirtual ? Twine("virtual ~") : Twine("~")) + StructureName);
```
Comment (lukasm91) [inline src/dawn/CodeGen/CXXUtil.h]:
```suggestion
                             (isVirtual ? Twine("virtual ~") : Twine("~")) + StructureName);
```

File: src/dawn/IIR/StencilInstantiation.cpp (hunk 3 above)
Comment (havogt) [reply src/dawn/IIR/StencilInstantiation.cpp]: Actually not. It is checking for count > 1, i.e. count >= 2.
I rewrote a bit, hope it's more readable now.

**Comment by havogt** [comment]
launch jenkins

**Pull request merged**